File: repository-hbase/src/main/java/hbase/util/RowKeyUtils.java
package hbase.util;

import model.AdminData;

import java.time.YearMonth;
import com.github.nscala_time.time.Imports.*;
import java.time.format.DateTimeFormatter;

public class RowKeyUtils {

    private static final String REFERENCE_PERIOD_FORMAT = "yyyyMM";
    static final String DELIMITER = "~";

    public static String getReferencePeriodFormat() {
        return REFERENCE_PERIOD_FORMAT;
    }


    public static String createRowKey(YearMonth referencePeriod, String id) {
        String period = referencePeriod.format(DateTimeFormatter.ofPattern(REFERENCE_PERIOD_FORMAT));
        return String.join(DELIMITER, period, id);
    }

    public static AdminData createAdminDataFromRowKey(String rowKey) {
        final String[] compositeRowKeyParts = rowKey.split(DELIMITER);
//        final YearMonth referencePeriod = YearMonth.parse(compositeRowKeyParts[0], DateTimeFormatter.ofPattern(REFERENCE_PERIOD_FORMAT));
        final YearMonth referencePeriod = YearMonth.parse(compositeRowKeyParts[0], DateTimeFormat.forPattern(REFERENCE_PERIOD_FORMAT)) ;
        /**
         * val referencePeriod: YearMonth =
         YearMonth.parse(compositeRowKeyParts.head, DateTimeFormat.forPattern(REFERENCE_PERIOD_FORMAT))
         */
        final String id = compositeRowKeyParts[1];
        return new AdminData(referencePeriod, id);
    }
}

File: repository-hbase/src/main/scala/hbase/util/RowKeyUtils.scala
package hbase.util

import javax.inject.Singleton

import com.github.nscala_time.time.Imports.{ DateTimeFormat, YearMonth }

import hbase.model.AdminData

/**
 * RowKeyUtils
 * ----------------
 * Author: haqa
 * Date: 03 November 2017 - 09:10
 * Copyright (c) 2017  Office for National Statistics
 */

@Singleton
class RowKeyUtils {

  val DELIMITER = "~"

  def reverseOption(id: String, flag: Boolean): String = if (flag) { id.reverse } else id

  def createRowKey(referencePeriod: YearMonth, id: String, flag: Boolean): String =
    String.join(DELIMITER, reverseOption(id, flag), referencePeriod.toString(AdminData.REFERENCE_PERIOD_FORMAT))

  def createAdminDataFromRowKey(rowKey: String, reverse: Boolean): AdminData = {
    val compositeRowKeyParts: Array[String] = rowKey.split(DELIMITER)
    val referencePeriod: YearMonth =
      YearMonth.parse(compositeRowKeyParts.last, DateTimeFormat.forPattern(AdminData.REFERENCE_PERIOD_FORMAT))
    val id = compositeRowKeyParts.head
    AdminData(referencePeriod, reverseOption(id, reverse))
  }

}
